### ISSUE TITLE
Usage Question

### ISSUE THREAD

**Issue by nickwhitman1993** [mfem/mfem#592]
Hi again,

I am looking to implement a Bilinear integrator and a Linear integrator with the weak form:

![codecogseqn 1](https://user-images.githubusercontent.com/23174133/44816707-e4182880-ab97-11e8-9a8d-33b2555e72cf.gif)

Where lambda is a VectorFunctionCoefficient and alpha is a FunctionCoefficient, and f is a known GridFunctionCoefficient.

Do you have any suggestions on where I might begin?

Thanks again, 

Nick

**Comment by mlstowell** [comment]
Hello, @nickwhitman1993,

The bilinear form can be computed using the existing DiffusionIntegrator with a MatrixCoefficient defined by alpha lambda lambda^T.  Fortunately writing coefficients is easier than BilinearFormIntegrators.  If you need some examples you can look in the branch `derived-coefs-dev` which I just created recently.

The linear form may be a little more difficult because we don't currently have any linear form integrators which integrate against derivatives of the basis functions.  You should be able to look at the DiffusionIntegrator to see how to work with gradient of the basis functions but you'll also need to look at DomainLFIntegrator to see how the linear form integrals are computed.

I can help you more with this if you are still unsure how to proceed.

Best wishes,
Mark

**Comment by nickwhitman1993** [comment]
Excellent thank you. I was using a MatrixCoefficient passed through a DiffusionIntegrator already, but looks like I do need to transpose it. I was getting unphysical results in some of the problem. Perhaps that was my error in what I am seeing. Is there a quick way to do that after I construct it?

Thanks again for your help! I'll see if I cant look into this issue and come back with more questions.

Best,

Nick

**Comment by mlstowell** [comment]
I believe this MatrixCoefficient is symmetric so you should not need to transpose the matrix.  The matrix I'm thinking of has entries defined by m_{ij} = alpha lambda_i lambda_j.

**Comment by nickwhitman1993** [comment]
So this is my current process:

![untitled](https://user-images.githubusercontent.com/23174133/44818445-ec736200-ab9d-11e8-91e9-608b88522432.jpg)

Do you see anything that jumps at you as incorrect? I've been racking my brain on this one. I am also benchmarking against another transport method so I know what I should be getting.

**Comment by nickwhitman1993** [comment]
Thank you. I fixed this now such that I define a function that instead of looping through, I just call it in to the problem: i.e.

double uniformInfMed::computeSigmaTinverse(mfem::Vector& x)
{
   // S0 = Q / (4*pi)
   return 1.0/this->computeSigmaT(x);
}

However, I still get some issues. Do you have any idea what might be going on by looking at the picture? 
The first is the "benchmark" and the second is what I am getting.
 
![visit0006](https://user-images.githubusercontent.com/23174133/44819992-db2d5400-aba3-11e8-8f01-1754fcce2a43.png)
![visit0007](https://user-images.githubusercontent.com/23174133/44819993-db2d5400-aba3-11e8-87d0-19d5245b3a41.png)

My guess is that I am missing some type of interior face integrator?

Additionally, there might be something in the boundary conditions. Unlike the first order form, I need to add in an additional term on the upwind problem boundaries:

**Comment by mlstowell** [comment]
An interior face integrator may come up with a DG method (I'm not expert on DG) but if you're using H1 finite elements you should not need this.

It could be a boundary condition issue as you suggest.  

I would recommend double checking your signs.  Note that MFEM's DiffusionIntegrator incorporates a negative sign because the weak form (Grad u, Grad v) is equivalent to (u, -Div(Grad v)).

**Comment by nickwhitman1993** [comment]
So perhaps it is the boundary conditions (maybe it means interior face boundaries). I need something like this then:

![gif](https://user-images.githubusercontent.com/23174133/44822082-d1105300-abad-11e8-8204-40533371ec01.gif)

Where I only add the term IF lambda dot normal is positive. Do you have any suggestions, then?

My thought was to use something similar to the DGTraceIntegrator and cleverly choose the coefficients such that I only get a term when it lambda dot normal is positive. But again, you said that I shouldn't need DG...?! 

Thanks again for all your help thus far, I really appreciate it.

Nick

**Comment by mlstowell** [comment]
Hi, Nick,
I know very little about DG methods.  If this is how you're formulating the problem then perhaps @v-dobrev or @YohannDudouit could offer you more useful advice.  If you're using H1 FiniteElements with the DiffusionIntegrator then you're not using DG (as I understand it).  If you do want to use DG then you should choose the DGDiffusionIntegrator (I cannot be much help beyond that however).
Best wishes,
Mark

**Comment by YohannDudouit** [comment]
Hi Nick,
Indeed more details on what you're trying to achieve, or just copy of a bigger part of your code so that I can check it from the beginning would be useful.

**Comment by nickwhitman1993** [comment]
Mark, thanks again for all your help!

Hi Yohann,

Thanks for helping! I will provide as many details as I can. Apologies as my background in finite elements is very lacking and I am new to all of this.

My colleague has already written a first-order discrete ordinates (Sn) neutron transport equation using DG mfem. You will see his snippet of code where it states "Sn mode". I am using his code to benchmark mine for comparison and understand what each weak form term does. 

Now for my portion: the basic premise of what I am tying to achieve is coding in the second-order self-adjoint angular flux (SAAF) discrete ordinates neutron transport equation, which in Galerkin weak form looks like this as the b(psi*,psi) = l(psi*) below.

![helpme](https://user-images.githubusercontent.com/23174133/44884403-15fcbe00-ac70-11e8-9071-02b8bc69ed10.png)

What I am currently doing is implementing the code in parts. The current test problem is driven externally on the left boundary by an isotropic scalar flux phi (psi = phi/4pi) and there is no scattering or internal drivers (think particle attenuation/shielding). This allows me to test if I have implemented the DiffusionIntegrator (second order term) correctly or not as well as the upwind terms before I start making a new Linear Intergrator for the (psi* + omega dot grad psi*, _____) term.

My next step is to indeed handle the scattering and internal source part in the linear form and test that next before moving on.

The code: SAAF 

     if (SAAFtrue == true)
      {
      //Project sigma_t to the grid in order to use it
      mfem::MatrixFunctionCoefficient omega_tensor(mat, sigma_t_inv); //TODO

      //Create tensor for -Grad dot OmegaOmega/sigma_t Grad psi
      mfem::BilinearFormIntegrator *tensor1 =
          new mfem::DiffusionIntegrator(omega_tensor);
      tensor1->SetIntRule(&elIntRule);
      L.AddDomainIntegrator(tensor1);

      // Upwind fluxes between zones on each face interior to mesh
      // For surface, we need a boundary geometry integration rule and order
      mfem::BilinearFormIntegrator *traceInterior =
          new mfem::DGTraceIntegrator(omega, -1.0, +0.5);
      traceInterior->SetIntRule(&bdrIntRule);

      // MFEM ISSUE: Maybe have the transpose operator pass along the setting of the
      // IntRule
      // to its owned worker.  Or at least throw an error.
      mfem::BilinearFormIntegrator *transInterior =
          new mfem::TransposeIntegrator(traceInterior);
      L.AddInteriorFaceIntegrator(transInterior);

      // Now add the inflow boundary conditions to the matrix and RHSs
      // need to SetIntRule on the DGTraceIntegrator rather than the
      // TransposeIntegrator
      mfem::BilinearFormIntegrator *traceBdr =
          new mfem::DGTraceIntegrator(omega, -1.0, +0.5);
      traceBdr->SetIntRule(&bdrIntRule);
      mfem::BilinearFormIntegrator *transBdr =
          new mfem::TransposeIntegrator(traceBdr);
      L.AddBdrFaceIntegrator(transBdr);


      mfem::LinearFormIntegrator *BndryFlow =
          new mfem::BoundaryFlowIntegrator(inflow, omega, -1.0, -0.5);
      BndryFlow->SetIntRule(&bdrIntRule);
      b.AddBdrFaceIntegrator(BndryFlow); // MFEM 3.2

    }
    else //------------------------------------------------------------------------
    {
      //SN mode
      // Add convection term within the zones
      mfem::BilinearFormIntegrator *convection =
          new mfem::ConvectionIntegrator(omega, +1.0);
      convection->SetIntRule(&elIntRule);
      L.AddDomainIntegrator(convection);
      //std::cout << "SN mode";

      // Upwind fluxes between zones on each face interior to mesh
      // For surface, we need a boundary geometry integration rule and order
      mfem::BilinearFormIntegrator *traceInterior =
          new mfem::DGTraceIntegrator(omega, -1.0, +0.5);
      traceInterior->SetIntRule(&bdrIntRule);

      // MFEM ISSUE: Maybe have the transpose operator pass along the setting of the
      // IntRule
      // to its owned worker.  Or at least throw an error.
      mfem::BilinearFormIntegrator *transInterior =
          new mfem::TransposeIntegrator(traceInterior);
      L.AddInteriorFaceIntegrator(transInterior);

      // Now add the inflow boundary conditions to the matrix and RHSs
      // need to SetIntRule on the DGTraceIntegrator rather than the
      // TransposeIntegrator
      mfem::BilinearFormIntegrator *traceBdr =
          new mfem::DGTraceIntegrator(omega, -1.0, +0.5);
      traceBdr->SetIntRule(&bdrIntRule);
      mfem::BilinearFormIntegrator *transBdr =
          new mfem::TransposeIntegrator(traceBdr);
      L.AddBdrFaceIntegrator(transBdr);

      //TODO This was Doug's original inflow conditions for the first-order equation.
       mfem::Array<int> in_bdr(4);
       in_bdr[0] = 1; // bottom
       in_bdr[1] = 1; // right
       in_bdr[2] = 1; // top
       in_bdr[3] = 1; // left
      mfem::LinearFormIntegrator *BndryFlow =
          new mfem::BoundaryFlowIntegrator(inflow, omega, -1.0, -0.5);
      BndryFlow->SetIntRule(&bdrIntRule);
      b.AddBdrFaceIntegrator(BndryFlow); // MFEM 3.2
      inflow_vis.ProjectCoefficient(inflow);
    }
        //END SN code -----------------------------------------------------------------------

    // SOURCE TERM //
    // TODO: Move this term outside the angle loop and pass it in. Only need to calc once per source iteration.
    mfem::FunctionCoefficient sigma_s(sigma_s_function);
    sigma_s_vis.ProjectCoefficient(sigma_s);
    for (int i = 0; i < phi_old.Size(); i++)
    {
      scat_src[i] = phi_old[i] * sigma_s_vis[i] / (4.0 * M_PI);
    }
    // Why are we taking the address of this?  It should take a reference.

    mfem::GridFunctionCoefficient flux(&scat_src);
    scat_src_vis.ProjectCoefficient(flux);

    mfem::LinearFormIntegrator *LFinteriorFlux =
        new mfem::DomainLFIntegrator(flux);
    LFinteriorFlux->SetIntRule(&elIntRule);
    b.AddDomainIntegrator(LFinteriorFlux);

     mfem::FunctionCoefficient qext(S0_function);
     qext_vis.ProjectCoefficient(qext);

    mfem::LinearFormIntegrator *LFinteriorQext =
        new mfem::DomainLFIntegrator(qext);
    LFinteriorQext->SetIntRule(&elIntRule);
    b.AddDomainIntegrator(LFinteriorQext);

Now I just ran some tests as the code currently sits. I used DG finite element collection for both tests (I tried using the H1 finite element collection when I switched to my code and it didn't change the solution much). What I did notice was that until I went to second-order finite elements, the solution of the second-order equation was NOT smooth, which retroactively makes more sense. Of course, the first-order equation is smooth with first-order elements.

The comparison of these two runs is below (both are with second-order finite elements):

Second-order equation:

![saaf12-dg-e20001](https://user-images.githubusercontent.com/23174133/44884883-b9e76900-ac72-11e8-8d10-78b4f536234f.jpeg)


First-order:

![s12-dg-e20002](https://user-images.githubusercontent.com/23174133/44884927-f87d2380-ac72-11e8-8b15-07319e000c57.jpeg)

Any thoughts or comments is greatly appreciated.

Best, 

Nick

**Comment by YohannDudouit** [comment]
Sorry for the delayed answer, I don't see anything that jumps to me as wrong. I don't see any reason why mfem wouldn't handle your formulation.
A few remarks:
- Do you really need to set the `IntegrationRule` yourself? Would the default one be incorrect? Are you sure you're not under-integrating with the `IntegrationRule` you give?
- Boundary conditions, and fluxes look "similar" in first order and second order to me. Could it be a problem of source initialization for the second order case? Are you sure you're not under-integrating your source term?
- Are you sure about the different signs in your flux terms? It can sometime be counter-intuitive to get the correct sign for average and jump terms.
- I don't see the "mass" (\psi*, \sigma \psi)_D term being added to your bilinearform
- Did you find a parameter that amplifies the result differences, _e.g._ mesh, order of the basis, order of the quadrature rule, etc?
- Did you try to refine the visualization to see if anything strange happens at the boundary of elements or of the domain?

**Comment by nickwhitman1993** [comment]
No problem, I appreciate your reply. 


- I probably don't need to set the integration rule myself. I am just using that as an artifact of inheriting the code. I can definitely omit that section and see what happens.

However, this portion was also included in the code (which I didn't include before). I'm not sure if this would answer the question you are asking. I believe that this code checks which order of FE and Mesh distortions I specify in at run-time and choose the integration order?

       int elementGeometry = fes.GetMesh()->GetElement(0)->GetGeometryType();
       // GetBdrElement(0) screams if there aren't actually any Bdrs (periodic mesh)
       int checkBE = fes.GetMesh()->GetNBE();
       //std::cout << "checkBE: " << checkBE << "\n";
       int borderElementGeometry;
       if(checkBE > 0)
       {
          borderElementGeometry =
             fes.GetMesh()->GetBdrElement(0)->GetGeometryType();
       }

      // From the guts of all the different integrators showing their defaults.
      // For Quads and hexes, less for triangles/tets.  Line numbers for mfem
      // version 6472008bca376a6083ccbdc02fb1ada07e14d0d1
      // k = mesh-order
      // d = element dimension
      // l = finite element order
      // OrderW = meshOrder * dim - 1                      fem/eltrans.cpp, 103
      // OrderGrad = meshOrder * (dim-1) + (feOrder-1)     fem/eltrans.cpp, 122
      // OrderJ = meshOrder                                fem/eltrans.cpp, 89
      // Order = meshOrder                                 fem/eltrans.hpp, 78
      // a = 2, b = 0 (user set)                           fem/lininteg.hpp, 54

      // mass: 2 * feOrder + OrderW                        fem/bilininteg.cpp, 501
      //           = 2 * feOrder + meshOrder * dim - 1
      // convection: OrderGrad + Order + feOrder           fem/bilininteg.cpp, 587
      //           = meshOrder * dim + 2 * feOrder - 1
      // trace: OrderW + 2 * feOrder                       fem/bilininteg.cpp, 1564
      //           = meshOrder * dim -1 + 2 * feOrder
      // boundary: OrderW + 2 * feOrder                    fem/lininteg.cpp, 392
      //           = meshOrder * dim - 1 + 2 * feOrder
      // domain(source): a * meshOrder + b                 fem/lininteg.cpp, 42
      //           = 2 * meshOrder
      //
      //
      // DIFFUSION OPERATORS
      //
      // DGDiffusion: 2 * feOrder                          fem/bilininteg.cpp
      // Diffusion: 2 * feOrder + dim - 1                  fem/bilininteg.cpp

      int meshOrder = fes.GetMesh()->GetElementTransformation(0)->Order();
      //int feOrder = fes.GetOrder(0);
      int elementIntegrationOrder = 0;
      if (feOrder == 0) {
        elementIntegrationOrder = 2 * meshOrder;
      }
      else if(meshOrder == 0) {
        elementIntegrationOrder = 2 * feOrder;
      } else {
        elementIntegrationOrder = 2 * (feOrder)+dim * meshOrder - 1;
      }
      int borderElementIntegrationOrder = elementIntegrationOrder;

      // MFEM ISSUE: Can we set this on the integrator?  Or even the bilinear or
      // linear forms?
      // Even better, specify elementIntegrationOrder, and let it get the rule for
      // the geometry that
      // exists, instead of assuming they are all the same on the mesh.
      const mfem::IntegrationRule &elIntRule =
          mfem::IntRules.Get(elementGeometry, elementIntegrationOrder);
       const mfem::IntegrationRule &bdrIntRule = mfem::IntRules.Get(elementGeometry, elementIntegrationOrder);
       if(checkBE > 0)
       {
          const mfem::IntegrationRule &bdrIntRule =
             mfem::IntRules.Get(borderElementGeometry, borderElementIntegrationOrder);
       }
- Boundary conditions do indeed appear similarly in the first and second order forms. The only difference is that the downwind boundaries need to be specified, but I believe this is achieved through the DGTraceIntegrator. Again, this isn't necessarily _needed_ in the first order form but was included due to a separation of the convection integrator into a domain integral and a surface integral using Green's theorem. Maybe not. Does this make sense? 
- I believe the signs are all correct.
- Correct - good catch. This term is actually included but I did not include it since it does not change between methods.
- I am working on looking at these items. It definitely seems that the lower the order of the quadrature rule the more difference. Which should make sense as well. 
- I'm not sure what you mean. I have not tried to do any AMR if that's what you mean.

I did run a few more tests and plotted using a logarithmic scale. I can run a few more tests after I remove the integration rule specification and see what happens. Given that the results are so similar, I am fairly convinced that I have implemented the integrators correctly. I can definitely change the integration rule if you think that it would affect the answer.

Additionally, it seems like the solutions only really behave physically if I provide either a 2nd order FE or 4th order FE on the second-order equation, and similarly for the first-order equation but with 1st order or 3rd order FE. This makes sense to you?

Second-Order, 2nd order FE, H1 FE collection
![saaf-h1-o2-eps0 10000](https://user-images.githubusercontent.com/23174133/45058259-fb9a5a00-b04c-11e8-9de4-b969107e25ba.png)

First-Order, 1st order FE, DG FE collection
![s12-dg-o1-eps0 10000](https://user-images.githubusercontent.com/23174133/45058270-01903b00-b04d-11e8-9096-022633e9c329.png)

Again, thanks for taking the time to sort through all my replies.

Best,

Nick

**Comment by YohannDudouit** [comment]
- I am very confused by the fact that you don't include the mass term, even if they are the same, your `CoefficientMatrix` is singular, could that result in a problem that is not well posed? _i.e._ several solutions exist and your different formulations don't converge to the same one?
- According to your plots, it looks like something wrong is happening with the boundary condition. Are you sure you're not enforcing some unneeded continuities by adding your boundary terms? Jumps and average enforces both normal and tangential continuity.
- The fact that the order of the quadrature rule has so much impact is suspicious, at some point it shouldn't make any difference.
- What I meant by refine the visualization is that because it is a high order discontinuous polynomial solution that you visualize by a continuous first order solution (I guess, is it VTK?), this might hide some strange behaviour at the boundaries. You could try using [GLVis](http://glvis.org/) to really see the solution jumps between elements, and sampling (refining) more with the 'i' keyboard input if needed. 'm' keyboard input can be useful to see your mesh too.

**Comment by nickwhitman1993** [comment]
- Sorry for the confusion. The mass term IS included I just did not include in the code snippet I sent you.
 Additionally, the solutions between the first-order and second-order do not necessarily HAVE to be the same answer. I just wanted to make sure they exhibit the same general behavior. I go into detail below but until I can compare against an analytical solution, I don't have an exact comparison. I simply am trying to make sure the diffusion integrator, tensor, and B.C.'s are implemented before I move on to the next task.
- I'm not exactly sure what you mean. The problem is supposed to be only driven on the left boundary. The other boundary conditions should be from the upwind terms. The boundary conditions are heterogeneous, and are 0 incident current (i.e. vacuum boundaries, besides the source).
- Perhaps I should clarify - the quadrature I am specifying is the order of the angular integral (i.e. the spatial discretization of the problem into weights and ordinates, Level-Symmetric) and is not related to anything in the finite-element space or finite-element related. 
- Additionally, this problem is a pure shielding problem, so I currently do not have an analytical solution I can compare to yet. This will be possible once I incorporate scattering into the problem and finish making the custom integrator for the ((1 + omega dot grad)u, f) portion. I will then use source-iteration to solve further.

-Nick

**Comment by YohannDudouit** [comment]
- What I meant by something looks wrong on your boundary condition (it might come from something else) is that your solution is `y` invariant `u(x,y) = u(x,0)` (it seems), which is not the case for the second plot.
- Looking at your second plot I also see a discontinuity in the bottom left corner(going from yellow to orange), it might be expected with DG methods, but it could also reveal a bug in your fluxes. If you expect your solution to be continuous, this jump should converge when refining.
- It seems ambitious to verify that your code (and method) is correct before you can compare it to something you know correct.

**Comment by nickwhitman1993** [comment]
- I don't believe it is completely y invariant. There is definitely y variation although it is difficult to see unless you are near the boundaries. The source on the boundary however, is indeed u(0, y) = u0

- The second plot is the first-order solution on a first-order FE. I think that for higher order FE this goes away

- I agree with you and I wasn't trying to imply that I was correct or the method is implemented correctly. However, I was merely trying to point out that the first-order method (which was not written by me) has been tested against analytical solutions and bench-marked already. This is why I use the first-order method as a comparison to make sure my code is correct as I add in features (which couldn't be tested independently unless I had no scattering and no internal source yet, which makes it impossible for an analytical solution). Hopefully that provides more context.

Apologies for coming across as if I was correct. I have no desire to make premature assumptions or false claims by any means until they are well founded.

-Nick

**Comment by nickwhitman1993** [comment]
@mlstowell I am in the process of trying to develop the linear form integrator for the derivative of the basis functions (i.e. VectorCoefficient dot Gradient Basis Function), and I am experiencing some difficulty.
Could you provide some guidance, perhaps? 

Nick

**Comment by mlstowell** [comment]
I'd be happy to.  Have you got a first draft we can work from?

**Comment by nickwhitman1993** [comment]
Thanks! I apologize in advance if it ends up being very rough.
Even if you could just point out a few mistakes so I can work on those, that would be greatly appreciated. 
This format does compile correctly while building mfem. However, when building my run directory I get an error  

error: expected type-specifier
         new mfem::DomainConvectionIntegrator(omega, flux);

Again, thanks so much!!

      class DomainConvectionIntegrator : public LinearFormIntegrator
      {
      private:
        DenseMatrix dshape, adjJ;
        Vector shape, vec2, BdFidxT, Q_ir;
        VectorCoefficient &Q;
        Coefficient &F;
        int oa, ob;
        double alpha;
      public:
        /// Constructs a domain integrator with a given Coefficient
        DomainConvectionIntegrator(VectorCoefficient &q, Coefficient &f, double a = 1.0)
           : Q(q), F(f) {}
         // public:
         //    /// Constructs a boundary integrator with a given Coefficient QG
         //    DomainConvectionIntegrator(VectorCoefficient &QG, int a = 1, int b = 1)
         //       : Q(QG), oa(a), ob(b) {};

            virtual void AssembleRHSElementVect(const FiniteElement &el,
                                                ElementTransformation &Tr,
                                                Vector &elvect);

            using LinearFormIntegrator::AssembleRHSElementVect;
      };

        void DomainConvectionIntegrator::AssembleRHSElementVect(const FiniteElement &el,
                                                        ElementTransformation &Tr,
                                                        Vector &elvect)
        {
          int nd = el.GetDof();
          int dim = el.GetDim();

          elvect.SetSize(nd);
          dshape.SetSize(nd,dim);
          adjJ.SetSize(dim);
          shape.SetSize(nd);
          vec2.SetSize(dim);
          BdFidxT.SetSize(nd);

          Vector vec1;

          const IntegrationRule *ir = IntRule;
          if (ir == NULL)
          {
             int order = Tr.OrderGrad(&el) + Tr.Order() + el.GetOrder();
             ir = &IntRules.Get(el.GetGeomType(), order);
          }

          elvect = 0.0;

          for (int i = 0; i < ir->GetNPoints(); i++)
          {
             const IntegrationPoint &ip = ir->IntPoint(i);
             el.CalcDShape(ip, dshape);
             el.CalcShape(ip, shape);

             Tr.SetIntPoint(&ip);
             CalcAdjugate(Tr.Jacobian(), adjJ);
             Q.Eval(Q_ir, Tr, ip); //VectorCoefficient omega
             adjJ.Mult(vec1, vec2);
             dshape.Mult(vec2, BdFidxT);

             double val = Tr.Weight() * F.Eval(Tr, ip) * BdFidxT[i]; //Coefficient

             elvect.Add(ip.weight * val, shape);
           }
        }

**Comment by YohannDudouit** [comment]
https://stackoverflow.com/questions/8845117/error-expected-type-specifier-before-classname

**Comment by YohannDudouit** [comment]
I think the problem comes from the code where you instantiate the class (not the code of class itself), could you make sure that none of what is described in the link is your problem?

**Comment by mlstowell** [comment]
Hi, @nickwhitman1993,

No problem, glad to help. 

A few initial points.  `Q_ir` does not seem to be used and `vec1` does not appear to be set.  Perhaps `Q_ir` should be used in place of `vec1`?  You should not require both `shape` and `dshape` in a linear for integrator because the basis should only appear once.  Let's try to parse the algorithm and see what should happen.

I believe you intend to integrate the product `F(x) * Q(x)\dot Grad phi(x)` where `F` is a scalar function, `Q` is a vector function, and `phi` is the basis function.  I think I would do this as follows:
```
      for (int i = 0; i < ir->GetNPoints(); i++)
      {
         // Grab the current integration point and pass it to the element transformation
         const IntegrationPoint &ip = ir->IntPoint(i);
         Tr.SetIntPoint(&ip);

         // Calculate the gradient of the basis functions at this point in physical space
         el.CalcPhysDShape(Tr, dshape);

         // Compute vector coefficient at this point
         Q.Eval(Q_ir, Tr, ip); //VectorCoefficient omega

         // Compute the dot product of Q with each basis function
         // and accumulate the results into the output vector with appropriate weight
         dshape.AddMult_a(ip.weight * Tr.Weight() * F.Eval(Tr, ip), Q_ir, elvect);
       }
```
Note that most of the Jacobian calculations are done inside the CalcPhysDShape method. This simplifies things a little bit by removing a couple of the temporary variables.

Let me know how this goes...

Mark

**Comment by mlstowell** [comment]
I agree with @YohannDudouit , the error message probably indicates a different issue.  Perhaps something with the namespace containing the new class (did you declare it in the `mfem` namespace?)

**Comment by nickwhitman1993** [comment]
Thanks a bunch @mlstowell and @YohannDudouit - Currently looking into that issue as well.

I'll give that a shot and see what I can do.

Nick

**Comment by nickwhitman1993** [comment]
@mlstowell Did you mean 
         // Calculate the gradient of the basis functions at this point in physical space
         el.CalcPhysDivShape(Tr, dshape);
?

**Comment by mlstowell** [comment]
No, `el` is a scalar basis isn't it?  Perhaps I misunderstood.  What basis are you using for your field in this case?

**Comment by mlstowell** [comment]
Hi, @nickwhitman1993 ,
Sorry, I didn't see your edit until just now.  This method is in the current master branch of MFEM but it may not have been in the most recent release.  This method is equivalent to the following but you'll need one temporary dense matrix:

```
   DenseMatrix vshape(nd, dim);
   ...
   CalcDShape(ip, vshape);
   Mult(vshape, Tr.InverseJacobian(), dshape);
```

**Comment by nickwhitman1993** [comment]
Thanks @mlstowell - It seems I am still having an issue with the type specifier. 

If I created the custom class in the lininteg.cpp and lininteg.hpp files, is there anywhere else I need to let the program know I have the new class?

**Comment by mlstowell** [comment]
No, but since this is a header file change you may need to delete and rebuild all of your object files.  Beyond that it should just be a matter of declaring the new class before using it.

**Comment by nickwhitman1993** [comment]
@mlstowell Thanks, that is what I thought as well. Must be another issue..

**Comment by mlstowell** [comment]
If you're willing to post a tar file I'd be happy to take a closer look.

**Comment by mlstowell** [comment]
Well, `LinearFormIntegrator` does have one pure virtual method which is `AssembleRHSElementVect`.  You are overriding this one so that shouldn't be a problem.  If there aren't any outdated object files floating around I'd check for small typos in the arguments of the overriding method.  It's possible the compiler doesn't recognize the overriding method if the arguments are even slightly different.

**Comment by nickwhitman1993** [comment]
Hi @mlstowell, for some reason when I recompiled, it didn't like the previous edits I had made. I figured out the error anyways. 

I realized that nowhere in my previous iteration of the linear integrator, let us call it the DomainConvectionIntegrator, did I multiply by my alpha FunctionCoefficient "alpha". I have now modified the class to accept another Coefficient in the argument.

At this point this works:
```
class DomainConvectionIntegrator : public LinearFormIntegrator
  {
  private:
    DenseMatrix dshape;
    Vector shape, Q_ir;
    VectorCoefficient &Q;
    Coefficient &S, &F;
  public:
    /// Constructs a domain integrator with a given Coefficient
    DomainConvectionIntegrator(VectorCoefficient &q, Coefficient &s, Coefficient &f,
      int a = 2, int b = 0)
       : Q(q), S(s), F(f) oa(a), ob(b) { }
     // public:
     //    /// Constructs a boundary integrator with a given Coefficient QG
     //    DomainConvectionIntegrator(VectorCoefficient &QG, int a = 1, int b = 1)
     //       : Q(QG), oa(a), ob(b) {};

        virtual void AssembleRHSElementVect(const FiniteElement &el,
                                            ElementTransformation &Tr,
                                            Vector &elvect);

        using LinearFormIntegrator::AssembleRHSElementVect;
  };

void DomainConvectionIntegrator::AssembleRHSElementVect(const FiniteElement &el,
                                                ElementTransformation &Tr,
                                                Vector &elvect)
{
  int nd = el.GetDof();
  int dim = el.GetDim();

  elvect.SetSize(nd);
  dshape.SetSize(nd,dim);
  shape.SetSize(nd);


  const IntegrationRule *ir = IntRule;
  if (ir == NULL)
  { 
     // This is the ingtegration rule I was using but I am not certain what I need... 

         int order = Tr.OrderGrad(&el) + Tr.Order() + el.GetOrder();
     ir = &IntRules.Get(el.GetGeomType(), order);
    
    // IR for DomainLFIntegrator
    // ir = &IntRules.Get(el.GetGeomType(),
    //                    oa * el.GetOrder() + ob + Tr.OrderW());
    //ir = &IntRules.Get(el.GetGeomType(), oa * el.GetOrder() + ob);
  }

  elvect = 0.0;


  for (int i = 0; i < ir->GetNPoints(); i++)
  {
     // Grab the current integration point and pass it to the element transformation
     const IntegrationPoint &ip = ir->IntPoint(i);
     Tr.SetIntPoint(&ip);

     // Calculate the gradient of the basis functions at this point in physical space
     el.CalcPhysDShape(Tr, dshape);

     // Compute vector coefficient at this point
     Q.Eval(Q_ir, Tr, ip); //VectorCoefficient omega

     // Compute the dot product of Q with each basis function
     // and accumulate the results into the output vector with appropriate weight
     // *This is where I multiplied alpha = S.Eval(Tr, ip)
     dshape.AddMult_a(ip.weight * Tr.Weight() * F.Eval(Tr, ip) * S.Eval(Tr,ip), Q_ir, elvect);
   }
}
```
But my current question is, did I implement the second portion correctly? Additionally, how would I know if I need to modify the integration rules for the `alpha lambda dot grad` portion? I am assuming the rules would be the same as the DomainLFIntegrator so I should change that if that is the case.

Basically I am splitting the Linear form into a DomainLFIntegrator and my custom integrator to handle it all. 
![codecogseqn 1](https://user-images.githubusercontent.com/23174133/44816707-e4182880-ab97-11e8-9a8d-33b2555e72cf.gif)

**Comment by mlstowell** [comment]
Hi, @nickwhitman1993 ,

Yes, it looks correct to me.

The appropriate integration rule order is a tricky question.  The `DomainLFIntegrator` in this case is only integrating the basis functions and its default integration order is probably sufficient to integrate these exactly (using a higher order there would be wasted effort).  If you know the polynomial orders of `alpha`, `lambda`, and `f` you could determine the appropriate integration order to integrate the `DomainConvectionIntegrator` exactly as well.  If even one of these is not a polynomial then you just need to choose something that you consider "good enough".

In my experience it can be important to use compatible integration orders for closely related weak forms (when they can't be exact).  This allows contributions from different integrals to cancel when they should.  I suspect that @YohannDudouit has more knowledge than I do regarding integration orders so I'll happily defer to him if he chooses to offer advice on this.  I can offer a few rules of thumb.  A bilinear form integrates four quantities; two basis functions (or their derivatives), the Jacobian, and the coefficient.  This suggests an integration order of `2*FEOrder+GeomOrder+CoefOrder` or, in the case of basis function derivatives `2*FEOrder-2+GeomOrder+CoefOrder`.  A compatible linear form integrator would use `FEOrder+GeomOrder+CoefOrder` or `FEOrder-1+GeomOrder+CoefOrder`.  Often `CoefOrder` is taken to be 0 or 1 although a higher value could be used.  It probably makes sense to restrict `CoefOrder<=FEOrder` because if this is not sufficient then you should probably refine the mesh to capture the small scale features of your coefficients.  It could also make sense to treat the function `f` as a separate factor in these considerations which would lead to `2*FEOrder+GeomOrder+CoefOrder` or `2*FEOrder-1+GeomOrder+CoefOrder`.  I realize this leaves a lot of room for uncertainty but I hope it is helpful.

Best wishes,
Mark
